https://bugs.gentoo.org/936594
From: Brahmajit Das <brahmajit.xyz@gmail.com>
Date: Tue, 30 Jul 2024 07:02:54 +0000
Subject: [PATCH 1/1] Fix implicit declaration of function basename on musl

On musl there is only one implementation of basename, which comes from
the libgen.h header file. However, in this case the code is using GNU
implementation of basename, which is not available on musl. This results
in build error:

ifrename.c:1816:15: error: implicit declaration of function basename [-Wimplicit-function-declaration]

Taking refernce from https://github.com/kmod-project/kmod/pull/32 we can
use a portable implementation of the basename API on musl (or globally).
But I'm open to better ideas.

Please also refer: https://bugs.gentoo.org/936594

Signed-off-by: Brahmajit Das <brahmajit.xyz@gmail.com>
--- a/ifrename.c
+++ b/ifrename.c
@@ -45,6 +45,7 @@
 #define _GNU_SOURCE
 #endif
 
+#include <string.h>
 #include <getopt.h>		/* getopt_long() */
 #include <linux/sockios.h>	/* SIOCSIFNAME */
 #include <fnmatch.h>		/* fnmatch() */
@@ -454,6 +455,18 @@ struct sysfs_metadata	sysfs_global =
   0, { NULL, NULL, NULL, NULL, NULL },
 };
 
+#if !defined(__GLIBC__)
+/*
+ * Use portable implementation for basename API
+ * https://github.com/kmod-project/kmod/pull/32
+ */
+char *gnu_basename(char *s)
+{
+  char *p = strrchr(s, '/');
+  return p ? p+1 : s;
+}
+#endif
+
 /******************** INTERFACE NAME MANAGEMENT ********************/
 /*
  * Bunch of low level function for managing interface names.
@@ -1813,7 +1826,11 @@ mapping_getsysfs(int			skfd,
 	  /* Here, we have a link name or a parent directory name */
 
 	  /* Keep only the last component of path name, save it */
+#if !defined(__GLIBC__)
+	  p = gnu_basename(linkpath);
+#else
 	  p = basename(linkpath);
+#endif
 	  sdup = strdup(p);
 	  free(linkpath);
 	}
-- 
2.45.2

